ntoskrnl/config/ntapi.c

```diff
@@ -1452,23 +1452,516 @@ NtLockRegistryKey(IN HANDLE KeyHandle)
+            PCM_POST_BLOCK SubPostBlock;
+            Status = CmpInsertSubPostBlock(SubNotifyBlocks[i],
+                                           NotifyBlock,
+                                           PostBlock,
+                                           &SubPostBlock);
+            if (!NT_SUCCESS(Status))
+            {
+                DPRINT1("NtNotifyChangeMultipleKeys: Failed to allocate and insert subordinate PostBlock. (0x%lx)\n", Status);
+                goto Failure2;
+            }
+            
+            /* The object is no longer needed */
+            CmpReleaseKcbLock(LocalSubObjectsKeyBody[i]->KeyControlBlock);
+            ObDereferenceObject(LocalSubObjectsKeyBody[i]);
+            LocalSubObjectsKeyBody[i] = NULL;
+
+            /* CmpInsertSubPostBlock adds a reference of PostBlock to NotifyBlock->PostList */
+        }
+
+        /* We don't need an array for storing KeyBody objects */
+        ExFreePoolWithTag(LocalSubObjectsKeyBody, TAG_CM);
+        LocalSubObjectsKeyBody = NULL;
+    }
+
+    /* Attach subordinate NotifyBlocks to the master's PostBlock */
+    PostBlock->SubCount = Count;
+    PostBlock->SubNotifyBlocks = SubNotifyBlocks;
+
+    /* We are done with the NotifyBlock, unlock the KCB so now user can change registry keys */
+    CmpReleaseKcbLock(KeyObject->KeyControlBlock);
+    CmpUnlockNotify();
+    KeLeaveCriticalRegion();
+
+    /* Initialize IO_STATUS_BLOCK */
+    if (!Asynchronous || PreviousMode != KernelMode)
+    {
+        IoStatusBlock->Status = STATUS_PENDING;
+        IoStatusBlock->Information = 0;
+    }
+
+    if (Asynchronous)
+    {
+        /* This is an asynchronous call, we set the notification up, return NOW and notify later */
+        /* The allocated resources will be freed later when notification session ended */
+        Status = STATUS_PENDING;
+        goto Cleanup;
+
+        /* FIXME: Update IoStatusBlock when a notification is sent, and report back the relative name of the changed key using the buffer */
+    }
+    else
+    {
+        /* Wait for event to be signaled */
+        KeWaitForSingleObject(PostBlock->Event, Executive, PreviousMode, FALSE, NULL);
+
+        /* FIXME: Report back the relative name of the changed using the Buffer */
+
+        /* PostBlock is freed automatically when the event is signaled */
+
+        Status = PostBlock->IoStatusBlock->Status;
+        goto Cleanup;
+    }
+
+Failure2:
+    CmpReleaseKcbLock(KeyObject->KeyControlBlock);
+
+    if (LocalEventObject)
+        ObDereferenceObject(LocalEventObject);
+
+    if (LocalEventHandle)
+        ZwClose(LocalEventHandle);
+
+    if (NotifyBlock)
+    {
+        if (KeyObject->NotifyBlock == NotifyBlock)
+        {
+            RemoveEntryList(&(NotifyBlock->HiveList));
+            KeyObject->NotifyBlock = NULL;
+        }
+
+        ExFreePoolWithTag(NotifyBlock, TAG_CM);
+    }
+
+    if (PostBlock)
+        ExFreePoolWithTag(PostBlock, TAG_CM);
+
+Failure:
```

Comment:
This is a lot of cleanup code, multiple exit paths, very hard to follow, if it is all correct. A sign, that you should refactor this function and split the inner logic into `CmpNotifyChangeKeys` or something like that.